### PR TITLE
Rescue from Psych::SyntaxError instead of SyntaxError after parsing YAML

### PR DESCRIPTION
**Pull request by jmhooper**
`SafeYAML` was raising a [`Psych::SynaxError`](https://github.com/ruby/psych/blob/master/lib/psych/syntax_error.rb), but the `Convertible` module and `Document` class were rescuing from a `SyntaxError`. This made the conditional branch for reporting that an error was a result of problems parsing the YAML unreachable and led to some more cryptic error messages.

```shell
Error reading file /Users/jonathanhooper/Developer/jekyll-test-site/pages/error.html: (<unknown>): did not find expected key while parsing a block mapping at line 2 column 1
```

versus

```shell
YAML Exception reading /Users/jonathanhooper/Developer/jekyll-test-site/pages/error.html: (<unknown>): did not find expected key while parsing a block mapping at line 2 column 1
```

This commit tells the `Document` class and `Convertible` module to rescue from the correct error class.

### CODE REVIEW

**Comment by jmhooper** [comment]
Note that I updated the tests for `Convertible.read_yaml`, but I was unable to find a matching test case for `Document.read`. I considered writing one, but had trouble finding where `Document.read` was used in the code and as such was uncertain about what a test for that would look like.

File: lib/jekyll/convertible.rb [hunk 1]
```diff
@@ -46,7 +46,7 @@ def read_yaml(base, name, opts = {})
           self.content = $POSTMATCH
           self.data = SafeYAML.load(Regexp.last_match(1))
         end
-      rescue SyntaxError => e
+      rescue Psych::SyntaxError => e
```
Comment (parkr) [inline lib/jekyll/convertible.rb]: This is very old code – can you leave the old `SyntaxError` and just add the `Psych::SyntaxError`?

**Comment by jmhooper** [comment]
I added this change in #5832 to avoid the inevitable conflict that would arise since it touches the same code. I move that this be closed if the change is accepted there.

**Comment by parkr** [comment]
Hey @jmhooper! We try to keep pull requests small and focused, so I'm going to have to ask you to make the necessary changes here so this change can be handled separately from your other change. This will ensure that this PR has a prominent listing in our changelog so our users can know what's up when it's shipped. Sorry about the conflict between this and that PR; we can always manually merge if it's too much of an inconvenience for you. Does that sound ok?

**Comment by jmhooper** [comment]
Yep, totally understand! I amended the commit on the other branch.

**Comment by ashmaroli** [comment]
@jmhooper please resolve the conflicts if you wish to further this PR. Thanks.

**Comment by ashmaroli** [comment]
```
lib/jekyll/document.rb:269:9: W: Lint/ShadowedException: Do not shadow rescued Exceptions.
        rescue Psych::SyntaxError, StandardError, Errors::FatalException => e
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
is most probably because **both `Psych::SyntaxError` and `Errors::FatalException`** are subclasses of `RuntimeError`. So, in theory, removing `Errors::FatalException` should appease RuboCop..

**Comment by ashmaroli** [comment]
@jmhooper Requesting a minor modification on this branch.. and this is good to go.. :smiley: 

1. `git pull origin jmhooper-psych-syntax-error` to update your local branch
2. Remove `StandardError` in ln#269 of `document.rb`
3. Commit and push to remote.

**Comment by jmhooper** [comment]
@ashmaroli: Done. Thanks ya'll for picking this up

**Comment by ashmaroli** [comment]
Uh-oh..This does not look good..
`StandardError` is required to let `ArgumentError` be rescued..
Sorry for the misinformation.. Lets bring `StandardError` back and temporarily disable RuboCop here..

The following should patch this..
```diff
@@ -254,25 +254,27 @@
     # Read in the file and assign the content and data based on the file contents.
     # Merge the frontmatter of the file with the frontmatter default
     # values
     #
     # Returns nothing.
+    # rubocop:disable Lint/ShadowedException
     def read(opts = {})
       Jekyll.logger.debug "Reading:", relative_path
 
       if yaml_file?
         @data = SafeYAML.load_file(path)
       else
         begin
           merge_defaults
           read_content(opts)
           read_post_data
-        rescue Psych::SyntaxError, Errors::FatalException => e
+        rescue Psych::SyntaxError, Errors::FatalException, StandardError => e
           handle_read_error(e)
         end
       end
     end
+    # rubocop:enable Lint/ShadowedException
 
     # Create a Liquid-understandable version of this Document.
     #
     # Returns a Hash representing this Document's data.
```

**Comment by ashmaroli** [comment]
@jmhooper Sorry for the misinformation, `StandardError` needs to be rescued to rescue `ArgumentError` as well.. So lets just disable Rubocop here temporarily

```diff
@@ -254,25 +254,27 @@
     # Read in the file and assign the content and data based on the file contents.
     # Merge the frontmatter of the file with the frontmatter default
     # values
     #
     # Returns nothing.
+    # rubocop:disable Lint/ShadowedException
     def read(opts = {})
       Jekyll.logger.debug "Reading:", relative_path
 
       if yaml_file?
         @data = SafeYAML.load_file(path)
       else
         begin
           merge_defaults
           read_content(opts)
           read_post_data
-        rescue Psych::SyntaxError, Errors::FatalException => e
+        rescue Psych::SyntaxError, Errors::FatalException, StandardError => e
           handle_read_error(e)
         end
       end
     end
+    # rubocop:enable Lint/ShadowedException
```

**Comment by jmhooper** [comment]
@ashmaroli: Quick question. Since `Psych::SyntaxError` and `Jekyll::Errors::FatalException` are subclasses of `StandardError`, wouldn't it make more sense to just rescue from `StandardError` and let `handle_read_error` sort out what is a subclass of what? I.e.:

```diff
     def read(opts = {})
       Jekyll.logger.debug "Reading:", relative_path
 
       if yaml_file?
         @data = SafeYAML.load_file(path)
       else
         begin
           merge_defaults
           read_content(opts)
           read_post_data
-        rescue Psych::SyntaxError, Errors::FatalException => e
+        rescue StandardError => e
           handle_read_error(e)
         end
       end
     end
```

That seems to solve the problem without having to disable a cop.

**Comment by ashmaroli** [comment]
rescue-ing only `StandardError` is going to fail tests which was "patched" in #6520 ..

**Comment by ashmaroli** [comment]
> Since `Psych::SyntaxError` and `Jekyll::Errors::FatalException` are subclasses of `StandardError`, wouldn't it make more sense to just rescue from `StandardError` and let `handle_read_error` sort out what..

You may be correct.. Tests pass locally.
Lets start a new PR based of `master`.. this branch has been dirtied much..

**Comment by jmhooper** [comment]
@ashmaroli: I went ahead and pushed the code to rescue from just `StandardError` so we can see what CI does.

**Comment by jmhooper** [comment]
...but happy to open another branch if that is what we want to do. If we're worried about a dirty history, we could just rebase it though.

**Comment by ashmaroli** [comment]
> If we're worried about a dirty history, we could just rebase it though.

No worries.. it'll be handled when the PR is being merged..
Thanks for your time and patience.. :grin:

**Comment by jmhooper** [comment]
@ashmaroli Looks like all the builds passed except for one of the appveyor ones that failed to install cucumber

**Comment by DirtyF** [comment]
@jekyllbot: merge +dev

**Pull request merged**